packages/kbn-alerting-types/action_variable.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+export interface ActionVariable {
+  name: string;
+  description: string;
+  deprecated?: boolean;
+  useWithTripleBracesInTemplates?: boolean;
+  usesPublicBaseUrl?: boolean;
+}
+
+/**
+ *  Returns all flattened keys from a deeply nested object as union
+ */
+type FlattenKeys<T extends Record<string, unknown>, Key = keyof T> = Key extends string
```

Comment:
Im personally not a huge fan of these complicated derived types. Theyre are difficult to understand and doesn't immediately tell the viewer what the resulting type is. Can we do this without deriving the types? I'd rather we are a bit more verbose but can be easily understood.

Reply:
We need a way for TS to warn us if someone adds a new context variable in `x-pack/plugins/alerting/server/task_runner/transform_action_params.ts`. Because the UI uses a flattened version of the context variables I created this type to convert the backend types to the UI types and get an error if something new is added. I would change the UI structure but did not want to change that much of the existing code. I am open to other ideas. Any suggestion?

Reply:
what if we hardcode the `ActionContextVariablesFlatten` properties? Maybe too much maintenance in the future?

Reply:
If we hardcode the `ActionContextVariablesFlatten` how will you get notified by TS if someone adds a new property in `ActionContextVariables`? We need an automated way where the moment I add a new context variable on the backend or the fronted it cascades to the whole codebase without having to remember to do it manually.